src/Mod/Sketcher/Gui/Command.cpp

```diff
@@ -1213,6 +1191,250 @@ bool CmdSketcherGrid::isActive()
+                auto checked = checkbox->checkState() == Qt::Checked;
+
+                if (value != checked) {
+                    const QSignalBlocker blocker(checkbox);
+                    checkbox->setChecked(value);
+                }
+            };
+
+            ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+
+            updateCheckBox(snapToObjects, hGrp->GetBool("SnapToObjects", true));
+
+            updateCheckBox(snapToGrid, hGrp->GetBool("SnapToGrid", false));
+
+            snapAngle->setValue(hGrp->GetFloat("SnapAngle", 5.0));
+
+            bool snapActivated = hGrp->GetBool("Snap", true);
+            snapToObjects->setEnabled(snapActivated);
+            snapToGrid->setEnabled(snapActivated);
+            angleLabel->setEnabled(snapActivated);
+            snapAngle->setEnabled(snapActivated);
+        }
+    }
+
+    void languageChange()
+    {
+
+        snapToObjects->setText(tr("Snap to objects"));
+        snapToObjects->setToolTip(tr("New points will snap to the currently preselected object. It will also snap to the middle of lines and arcs."));
+        snapToObjects->setStatusTip(snapToObjects->toolTip());
+
+        snapToGrid->setText(tr("Snap to Grid"));
+        snapToGrid->setToolTip(tr("New points will snap to the nearest grid line.\nPoints must be set closer than a fifth of the grid spacing to a grid line to snap."));
+        snapToGrid->setStatusTip(snapToGrid->toolTip());
+
+        angleLabel->setText(tr("Snap angle"));
+        snapAngle->setToolTip(tr("Angular step for tools that use 'Snap at Angle' (line for instance). Hold CTRL to enable 'Snap at Angle'. The angle start from the East axis (horizontal right)"));
+    }
+
+protected:
+    QWidget* createWidget(QWidget* parent) override
+    {
+        snapToObjects = new QCheckBox();
+
+        snapToGrid = new QCheckBox();
+
+        angleLabel = new QLabel();
+
+        snapAngle = new Gui::QuantitySpinBox();
+        snapAngle->setProperty("unit", QVariant(QStringLiteral("deg")));
+        snapAngle->setObjectName(QStringLiteral("snapAngle"));
+        snapAngle->setMaximum(99999999.0);
+        snapAngle->setMinimum(0);
+
+        QWidget* snapW = new QWidget(parent);
+        auto* layout = new QGridLayout(snapW);
+        layout->addWidget(snapToGrid, 0, 0, 1, 2);
+        layout->addWidget(snapToObjects, 1, 0, 1, 2);
+        layout->addWidget(angleLabel, 2, 0);
+        layout->addWidget(snapAngle, 2, 1);
+
+        languageChange();
+
+        QObject::connect(snapToObjects, &QCheckBox::stateChanged, [this](int state) {
+            auto* sketchView = getView();
+
+            if (sketchView) {
+                ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+                hGrp->SetBool("SnapToObjects", state == Qt::Checked);

```

Comment:
Same, sketchView is there, unused. Then preference parameters used for communication.

Reply:
I don't understand why you say that. It's not just communication, we want to remember the user preferences. Else the settings are going to be reset everytime someone opens a sketch.

Reply:
When entities can be accessed directly via an API, generally no indirect paths are used. There are two reasons for it: (1) there is a performance impact associated, (2) the code is more difficult to follow and thus to debug. If it was merely about communicating, using parameters would look like a hack.

However, my remark here was not accurate as in this case we want to keep the state of the tool, including over different executions of FreeCAD. Then, it starts making sense again.

When we rely on a given mechanism (parameters) for persistence of the tool state, then communication needs to be reviewed. So what initially was an implementation that had drawbacks, is actually what needs to be used (albeit, taking a little bit of care about when to update, specially polling is costly with parameters, that is why we use observers, which is kind of event/interrupt driven).

I got up early today and decided to fix the issues to try to move forward with this PR. Please look at my commits, I have separated them on purpose so that you can easily follow what was the issue I perceived and what is the solution I provided. Of course, come back if you have questions or think anything is not right.